express_relay/sdk/python/express_relay_utils/express_relay_client.py

```diff
@@ -0,0 +1,404 @@
+        ) = None,
+    ):
+        """
+        Continually handles new ws messages as they are received from the server via websocket.
+
+        Args:
+            opportunity_callback: An async function that serves as the callback on a new opportunity. Should take in one external argument of type Opportunity.
+            bid_status_callback: An async function that serves as the callback on a new bid status update. Should take in one external argument of type BidStatusUpdate.
+        """
+        if not self.ws:
+            raise ExpressRelayClientException("Websocket not connected")
+
+        async for msg in self.ws:
+            msg_json = json.loads(msg)
+
+            if msg_json.get("type"):
+                if msg_json.get("type") == "new_opportunity":
+                    if opportunity_callback is not None:
+                        opportunity = Opportunity.model_validate(
+                            msg_json.get("opportunity")
+                        )
+                        if opportunity:
+                            asyncio.create_task(opportunity_callback(opportunity))
+
+                elif msg_json.get("type") == "bid_status_update":
+                    if bid_status_callback is not None:
+                        id = msg_json.get("status").get("id")
+                        bid_status = (
+                            msg_json.get("status").get("bid_status").get("status")
+                        )
+                        result = msg_json.get("status").get("bid_status").get("result")
+                        bid_status_update = BidStatusUpdate(
+                            id=id, bid_status=BidStatus(bid_status), result=result
+                        )
+                        asyncio.create_task(bid_status_callback(bid_status_update))
+
+            elif msg_json.get("id"):
+                future = self.ws_msg_futures.pop(msg_json["id"])
+                future.set_result(msg_json)
+
+    async def get_opportunities(
+        self, chain_id: str | None = None, timeout_secs: int = 10
+    ) -> list[Opportunity]:
+        """
+        Connects to the server and fetches opportunities.
+
+        Args:
+            chain_id: The chain ID to fetch opportunities for. If None, fetches opportunities across all chains.
+            timeout_secs: The timeout for the HTTP request in seconds.
+        Returns:
+            A list of opportunities.
+        """
+        params = {}
+        if chain_id:
+            params["chain_id"] = chain_id
+
+        async with httpx.AsyncClient() as client:
+            resp = await client.get(
+                urllib.parse.urlparse(self.server_url)
+                ._replace(path="/v1/opportunities")
+                .geturl(),
+                params=params,
+                timeout=timeout_secs,
+            )
+
+        resp.raise_for_status()
+
+        opportunities = [
+            Opportunity.model_validate(opportunity) for opportunity in resp.json()
```

Comment:
I think here is a better place to check the version before passing it to model_validate. Or we can catch the incompatible version errors here